src/Compilers/Core/Portable/CodeGen/TokenMap.cs

```diff
@@ -24,74 +23,91 @@ namespace Microsoft.CodeAnalysis.CodeGen
+        internal TokenMap()
+        {
+        }
 
-        internal TokenMap(IEqualityComparer<object> comparer)
+        public uint GetOrAddTokenFor(IReference item, out bool referenceAdded)
         {
-            _itemToToken = new Dictionary<object, uint>(comparer);
+            if (_itemIdentityToToken.TryGetValue(new IReferenceOrISignature(item), out uint token))
+            {
+                referenceAdded = false;
+                return token;
+            }
+
+            return AddItem(new IReferenceOrISignatureEquivalent(item), out referenceAdded);
         }
 
-        public uint GetOrAddTokenFor(object item, out bool referenceAdded)
+        public uint GetOrAddTokenFor(ISignature item, out bool referenceAdded)
         {
-            uint tmp;
-            if (_itemIdentityToToken.TryGetValue(item, out tmp))
+            if (_itemIdentityToToken.TryGetValue(new IReferenceOrISignature(item), out uint token))
             {
                 referenceAdded = false;
-                return (uint)tmp;
+                return token;
             }
 
-            return AddItem(item, out referenceAdded);
+            return AddItem(new IReferenceOrISignatureEquivalent(item), out referenceAdded);
         }
 
-        private uint AddItem(object item, out bool referenceAdded)
+        private uint AddItem(IReferenceOrISignatureEquivalent item, out bool referenceAdded)
         {
-            Debug.Assert(item is Cci.ISignature || item is Cci.IReference);
             uint token;
-
             // NOTE: cannot use GetOrAdd here since items and itemToToken must be in sync
             // so if we do need to add we have to take a lock and modify both collections.
-            lock (_items)
+            lock (_itemEquivalentToToken)
             {
-                if (!_itemToToken.TryGetValue(item, out token))
+                // Check if there is an equivalent type that has a token
+                if (!_itemEquivalentToToken.TryGetValue(item, out token))
                 {
-                    token = (uint)_items.Count;
-                    _items.Add(item);
-                    _itemToToken.Add(item, token);
+                    token = (uint)_count;
+                    // No equivalent, add the token for this type
+                    _itemEquivalentToToken.Add(item, token);
+
+                    var count = (int)token + 1;
+                    var items = _items;
+                    if (items.Length > count)
+                    {
+                        items[(int)token] = item.AsObject();
+                    }
+                    else
+                    {
+                        // Not enough room, we need to resize the array
+                        Array.Resize(ref items, Math.Max(8, count * 2));
+                        items[(int)token] = item.AsObject();
+
+                        // Update the updated array reference before updating _count
+                        Volatile.Write(ref _items, items);
```

Comment:
Why is this safe? It seems very likely to me that _items could have been changed by this point.

Reply:
This is the only location `_items` is updated and it is done in a lock. The `Volatile.Write` is to ensure its not reordered with the write to `_count`; so it's safe to `Volatile.Read` them in reverse order in `GetAllItems()` since `_count` only ever increases